sky/shell/ui/animator.cc

```diff
@@ -7,151 +7,102 @@
-    TRACE_EVENT_INSTANT1("flutter", "Frame request deferred",
-                         TRACE_EVENT_SCOPE_THREAD, "outstanding_requests_",
-                         outstanding_requests_);
-    did_defer_frame_request_ = true;
-    return;
-  }
-
-  if (!AwaitVSync()) {
-    base::MessageLoop::current()->PostDelayedTask(
-        FROM_HERE,
-        base::Bind(&Animator::BeginFrame, weak_factory_.GetWeakPtr(), 0),
-        base::TimeDelta::FromMilliseconds(16));
-  }
-}
+Animator::~Animator() = default;
 
 void Animator::Stop() {
   paused_ = true;
 }
 
 void Animator::Start() {
-  Reset();
-  RequestFrame();
-}
+  if (!paused_) {
+    return;
+  }
 
-void Animator::Animate(mojo::gfx::composition::FrameInfoPtr frame_info) {
-  BeginFrame(frame_info->frame_time);
+  paused_ = false;
+  RequestFrame();
 }
 
 void Animator::BeginFrame(int64_t time_stamp) {
-  TRACE_EVENT_ASYNC_END0("flutter", "Frame request pending", this);
-  DCHECK(engine_requested_frame_);
-  DCHECK(outstanding_requests_ > 0);
-  DCHECK(outstanding_requests_ <= kPipelineDepth) << outstanding_requests_;
-
-  engine_requested_frame_ = false;
-
-  if (paused_) {
-    OnFrameComplete();
+  pending_frame_semaphore_.Signal();
+
+  LayerTreePipeline::Producer producer = [this]() {
+    renderable_tree_.reset();
+    engine_->BeginFrame(ftl::TimePoint::Now());
+    return std::move(renderable_tree_);
+  };
+
+  if (!layer_tree_pipeline_->Produce(producer)) {
+    TRACE_EVENT_INSTANT0("flutter", "ConsumerSlowDefer",
+                         TRACE_EVENT_SCOPE_PROCESS);
+    RequestFrame();
     return;
   }
 
-  begin_time_ = ftl::TimePoint::Now();
-  ftl::TimePoint frame_time =
-      time_stamp
-          ? ftl::TimePoint() + ftl::TimeDelta::FromMicroseconds(time_stamp)
-          : begin_time_;
-
-  is_ready_to_draw_ = true;
-  engine_->BeginFrame(frame_time);
-  bool was_ready_to_draw = is_ready_to_draw_;
-  is_ready_to_draw_ = false;
-
-  // If we were still ready to draw when done with the frame, that means we
-  // didn't draw anything this frame and we should acknowledge the frame
-  // ourselves instead of waiting for the rasterizer to acknowledge it.
-  if (was_ready_to_draw)
-    OnFrameComplete();
+  auto weak_rasterizer = rasterizer_->GetWeakRasterizerPtr();
+  auto pipeline = layer_tree_pipeline_;
+
+  blink::Threads::Gpu()->PostTask([weak_rasterizer, pipeline]() {
+    if (!weak_rasterizer) {
+      return;
+    }
+    weak_rasterizer->Draw(pipeline);
+  });
 }
 
 void Animator::Render(std::unique_ptr<flow::LayerTree> layer_tree) {
-  if (!is_ready_to_draw_)
-    return;  // Only draw once per frame.
-  is_ready_to_draw_ = false;
-
-  layer_tree->set_construction_time(ftl::TimePoint::Now() - begin_time_);
-
-  // TODO(abarth): Doesn't this leak if OnFrameComplete never runs?
-  rasterizer_->Draw(
-      reinterpret_cast<uint64_t>(layer_tree.release()),
-      base::Bind(&Animator::OnFrameComplete, weak_factory_.GetWeakPtr()));
+  renderable_tree_ = std::move(layer_tree);
```

Comment:
This approach breaks an optimization we had previously.  Consider the following timeline:

* BeginFrame
* ... work to compute the new layer tree
* Render
* ... more work on this call stack before returning to BeginFrame

Previously the call to Render would kick the layer tree over to the GPU thread, which could begin processing it immediately.  Now we'll wait for the additional work that happens after the call to Render but before the call stack unwinds back to BeginFrame.

In practice, this isn't such a large optimization, but there's other code in the system that assumes that work it does after calling Render isn't holding up the frame.  It would be ok to land this patch as-is and address this issue in a follow-up patch, but the ideal patch here wouldn't lose this optimization.